 9. We will now consider the Boston housing data set, from the ISLP
 library.

 (a) Based on this data set, provide an estimate for the population
 mean of medv. Call this estimate ˆµ.


In [1]:
from ISLP import load_data
from IPython.display import display, Math
import numpy as np

boston = load_data('Boston')

# Estimate the population mean of `medv`
mu_hat = boston['medv'].mean()

# Display mu_hat in LaTeX with two decimal places
display(Math(f"\\hat{{\\mu}} = {mu_hat:.2f}"))

<IPython.core.display.Math object>

(b) Provide an estimate of the standard error of ˆµ. Interpret this
 result.
 
 Hint: We can compute the standard error of the sample mean by
 dividing the sample standard deviation by the square root of the
 number of observations.

In [2]:
n = len(boston['medv'])                  
s = boston['medv'].std(ddof=1)           
se_formula = s / np.sqrt(n)    

print(f'n = {n}')
print(f's (sample sd, ddof=1) = {s:.4f}')
print(f'se_formula (s/sqrt(n)) = {se_formula:.4f}')

n = 506
s (sample sd, ddof=1) = 9.1971
se_formula (s/sqrt(n)) = 0.4089


 (c) Now estimate the standard error of ˆµ using the bootstrap. How
 does this compare to your answer from (b)?

In [3]:
def bootstrap_se(data, n_boot=1000):
    boot_means = []
    for _ in range(n_boot):
        boot_sample = np.random.choice(data, size=len(data), replace=True)
        boot_means.append(np.mean(boot_sample))
    return np.std(boot_means, ddof=1)

# Calculate bootstrap SE for mean
se_boot_mean = bootstrap_se(boston['medv'])

# Recalculate formula SE for comparison
n = len(boston['medv'])
s = boston['medv'].std(ddof=1)
se_formula = s / np.sqrt(n)

print(f'se_boot_mean (bootstrap) = {se_boot_mean:.4f}')
print('Comparison: formula SE vs bootstrap SE:')
print(f'  formula SE = {se_formula:.4f}')
print(f'  bootstrap SE = {se_boot_mean:.4f}')
print(f'  Difference = {abs(se_formula - se_boot_mean):.4f}')

se_boot_mean (bootstrap) = 0.4167
Comparison: formula SE vs bootstrap SE:
  formula SE = 0.4089
  bootstrap SE = 0.4167
  Difference = 0.0078


 (d) Based on your bootstrap estimate from (c), provide a 95% confidence interval for the mean of medv. Compare it to the results obtained by using Boston['medv'].std() and the two standard error rule.

 Hint: You can approximate a 95% confidence interval using the
 formula 
 
 [ˆµ − 2SE(ˆµ), ˆµ + 2SE(ˆµ)].

In [8]:
# (d) 95% CI for mean

# Recalculate necessary values
mu_hat = boston['medv'].mean()
n = len(boston['medv'])
s = boston['medv'].std(ddof=1)
se_formula = s / np.sqrt(n)

def bootstrap_se(data, n_boot=1000):
    boot_means = []
    for _ in range(n_boot):
        boot_sample = np.random.choice(data, size=len(data), replace=True)
        boot_means.append(np.mean(boot_sample))
    return np.std(boot_means, ddof=1)

# Calculate bootstrap SE for mean
se_boot_mean = bootstrap_se(boston['medv'])

# 2*bootstrap SE rule
ci_boot_mean_2se = [mu_hat - 2 * se_boot_mean, mu_hat + 2 * se_boot_mean]

# 2*formula SE rule
ci_formula_2se = [mu_hat - 2 * se_formula, mu_hat + 2 * se_formula]

print('95% CI (2*bootstrap SE rule) =', ci_boot_mean_2se)
print('95% CI (2*formula SE rule) =', ci_formula_2se)

95% CI (2*bootstrap SE rule) = [np.float64(21.668859602871382), np.float64(23.39675304534997)]
95% CI (2*formula SE rule) = [np.float64(21.715084029115605), np.float64(23.35052861910575)]


 (e) Based on this data set, provide an estimate, ˆµmed, for the median
 value of medv in the population.


In [9]:
# (e) sample median
median_hat = boston['medv'].median()
print(f'median_hat = {median_hat}')

median_hat = 21.2


 (f) We now would like to estimate the standard error of ˆµmed. Unfortunately, there is no simple formula for computing the standard error of the median. Instead, estimate the standard error of the
 median using the bootstrap. Comment on your findings.

In [10]:
# (f) bootstrap SE for median
def bootstrap_se_median(data, n_boot=1000):
    boot_medians = []
    for _ in range(n_boot):
        boot_sample = np.random.choice(data, size=len(data), replace=True)
        boot_medians.append(np.median(boot_sample))
    return np.std(boot_medians, ddof=1)

se_boot_median = bootstrap_se_median(boston['medv'])
print(f'se_boot_median = {se_boot_median}')

se_boot_median = 0.37113922161612295


The bootstrap standard error for the median is 0.3711. This value indicates the variability of the median estimate across bootstrap samples. Compared to the mean's SE (around 0.4089), the median's SE is slightly smaller, which is expected as the median is more robust to outliers in skewed data like housing prices. Bootstrap is ideal here since no closed-form formula exists for the median's SE.

 (g) Based on this data set, provide an estimate for the tenth percentile of medv in Boston census tracts. Call this quantity ˆµ0.1.
 (You can use the np.percentile() function.)

In [ ]:
pct10_hat = np.percentile(boston['medv'], 10)
print(f'pct10_hat = {pct10_hat}')

pct10_hat = 12.75


 (h) Use the bootstrap to estimate the standard error of ˆµ0.1. Comment on your findings.

In [13]:
# (h) bootstrap SE for 10th percentile
def bootstrap_se_pct10(data, n_boot=1000):
    boot_pct10s = []
    for _ in range(n_boot):
        boot_sample = np.random.choice(data, size=len(data), replace=True)
        boot_pct10s.append(np.percentile(boot_sample, 10))
    return np.std(boot_pct10s, ddof=1)

se_boot_pct10 = bootstrap_se_pct10(boston['medv'])
print(f'se_boot_pct10 = {se_boot_pct10}')

se_boot_pct10 = 0.5103137246808062


The bootstrap standard error for the 10th percentile is 0.5103. This relatively large SE reflects the higher variability in estimating tail statistics like percentiles, as they depend on the extreme values in the data distribution. Compared to the mean (SE ~0.41) or median (SE ~0.37), the 10th percentile's SE is larger because small changes in the sample can significantly affect the lower tail. Bootstrap is essential here since no analytical formula exists for percentile SE, and it provides a reliable estimate for skewed data like housing prices.